mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -573,51 +574,88 @@ unsafe impl Sync for MetaRequest {}
 /// Future returned by `MetaRequest::write()`. It will complete when the write completes,
 /// or cancel the meta-request if dropped.
 #[derive(Debug)]
-pub struct MetaRequestWrite<'a> {
-    /// Signals when the write operation completes
-    future: FutureVoid,
+pub struct MetaRequestWrite<'r, 's> {
     /// The meta-request to cancel if this future is dropped
-    request: &'a mut MetaRequest,
+    request: &'r mut MetaRequest,
+    /// The slice to write
+    slice: &'s [u8],
+    /// Is end-of-file?
+    eof: bool,
+    /// Waker registered with `poll_write`
+    waker: Arc<Mutex<Option<Waker>>>,
 }
 
-impl Drop for MetaRequestWrite<'_> {
-    fn drop(&mut self) {
-        if !self.future.is_done() {
-            // This future is being dropped before completion. Cancelling the meta-request
-            // guarantees that the client will not access the provided buffer.
-            self.request.cancel();
+impl<'r, 's> MetaRequestWrite<'r, 's> {
+    fn new(request: &'r mut MetaRequest, slice: &'s [u8], eof: bool) -> Self {
+        Self {
+            request,
+            slice,
+            eof,
+            waker: Default::default(),
         }
     }
 }
 
-impl<'a> MetaRequestWrite<'a> {
-    fn new(request: &'a mut MetaRequest, slice: &'a [u8], eof: bool) -> Self {
-        // SAFETY: `MetaRequestWrite` will ensure that `slice` is alive until the future completes or
-        // that the meta-request is canceled if the future is dropped, preventing further use of the
-        // `aws_byte_cursor`.
-        let data = unsafe { slice.as_aws_byte_cursor() };
-
-        // SAFETY: `aws_s3_meta_request_write` never returns NULL.
-        let future = unsafe {
-            FutureVoid::from_crt(NonNull::new_unchecked(aws_s3_meta_request_write(
-                request.inner.as_ptr(),
+impl<'r, 's> Future for MetaRequestWrite<'r, 's> {
+    type Output = Result<&'s [u8], Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        let mut waker = self.waker.lock().unwrap();
+        if let Some(ref mut waker) = *waker {
+            // The write is still pending, just replace the waker.
```

Comment:
That's because in same cases it may change between `poll` invocations. See https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll:
> Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup.